enhancements/update/upgrades-blocking-on-ack.md

```diff
@@ -0,0 +1,210 @@
+
+* As an OCP product team we do not want customers to be surprised by workloads breaking due to APIs being removed after they upgrade.
+
+### Implementation Details/Notes/Constraints
+
+### Risks and Mitigations
+
+* A common risk is human error resulting in an incorrect key/value pair being placed in the `admin-acks` configmap.
+This type of error will not change `Upgradeable=false` and therefore the administrator will know a mistake has been made since the `Upgradeable` condition message will continue to be displayed.
+CVO will log an error whenever an unrecognized key is found in the `admin-acks` configmap.
+* Another risk is that the `admin-acks` configmap gets removed.
+This will result in `Upgradeable=false` and the standard `Upgradeable` condition message will be displayed with the additional message that the configmap cannot be found and should be created.
+CVO will recreate the `admin-gates` configmap if removed since it will be managed by CVO.
+CVO will log an error whenever the `admin-acks` configmap has been removed.
+* Unable to access the `admin-acks` configmap.
+This will result in `Upgradeable=false` and the standard `Upgradeable` condition message will be displayed with the additional message that the configmap cannot be accessed.
+CVO will log an error whenever there is an issue accessing the `admin-acks` configmap.
+* Unable to access the `admin-gates` configmap.
+Since CVO will not be able to determine what if any minor level upgrade blocks are defined CVO will set `Upgradeable=false` and set the displayed `Upgradeable` condition message to indicate that the `admin-gates` configmap cannot be accessed.
+CVO will log an error whenever there is an issue accessing the `admin-gates` configmap.
+
+## Design Details
+
+### Test Plan
+
+* CVO unit tests will be expanded and/or created to test the new logic.
+* e2e upgrade test will be enhanced.
+
+### Graduation Criteria
+
+GA. When it works, we ship it.
+
+#### Dev Preview -> Tech Preview
+
+N/A. This is not expected to be released as Dev Preview.
+
+#### Tech Preview -> GA
+
+N/A. This is not expected to be released as Tech Preview.
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+Inherent in the key name is an OCP major/minor version number. Based on this version number CVO knows to which OCP versions a given key applies.
+
+### Version Skew Strategy
+
+No special consideration.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation History`.
+
+## Drawbacks
+
+* Keys will not be portable across products that use a different versioning scheme than OCP.
+* `Upgradeable=false` is the default state and results if there is any issue accessing either configmap or if an invalid key is found in the `admin-acks` configmap.
+
+## Alternatives
+
+No alternatives have been discussed.
```

Comment:
Can we mention [the cloud-cred approach][1] ([`cloudcredential.openshift.io/upgradeable-to`][2])?  Because with this enhancement we're moving from "each component handles this directly and feeds the CVO via their ClusterOperator's `Upgradeable` condition" to "that's still allowed, but we're also providing this CVO-hosted UI", and I think it's worth discussing our motivation.  Probably:

> Admins may not want to go hunting through a bunch of component-specific UXes.  And mumble, mumble, about why browsing `ClusterOperator` `Upgradeable=False` condition messages wouldn't be enough to point admins at the steps they should be taking...

or something?

[1]: https://github.com/openshift/enhancements/pull/828#discussion_r670630101
[2]: https://github.com/openshift/cloud-credential-operator/pull/286/files#diff-f6d93feef6d59ca8ced14b54fe0058d41f14de44450baf6b723cc1c2ec280c54R132

Reply:
CCO's needs for admin signoff are ongoing (not one-offs). So we came up with a signoff where the admin says "i'm okay to upgrade to 4.X" (this should in theory allow upgrades from one EUS version to the next EUS version).
But it is odd for CCO to have its own mechanism that exists outside of this ConfigMap approach. Perhaps it is justified as CCO's needs are not release-specific, but ongoing pre-upgrade maintenance expected from the admin (and CCO's requirements about what is required from the admin is dependent on how CCO is configured to run eg. Manual/Mint/Passthrough/etc).

Reply:
> Admins may not want to go hunting through a bunch of component-specific UXes. And mumble, mumble, about why browsing ClusterOperator Upgradeable=False condition messages wouldn't be enough to point admins at the steps they should be taking...

The motivation in my mind is more that this isn't a specific component limitation/issue.  (You can argue that the apiserver should set upgradeable=false since the apiserver is removing the apis, but the apiserver isn't going to break if you upgrade to 4.9, its your workloads that are going to break)

but yes, agree that this shouldn't be setting a new precedent that the CVO manages upgradeable=false for everything.

Reply:
I don't think there's an action here?

I don't see this as setting a new precedent either. We're setting Upgradeable=false based on a cluster wide condition - the general case being we've asked the admin to do something and then ack that it's done.

Reply:
If there was an alternative discussed and rejected, it would be good to capture that info in the doc for completeness.

Reply:
the main alternative i can think of is we debated whether the apiserver should be the component setting upgradeable=false and looking for the ack, since it is the component that is "removing" the apis.

we rejected it because:
1) the apiserver is removing the apis, but the apiserver is not blocked from upgrading because of that..it would be more than happy to upgrade, regardless of what apis are in use.
2) this mechanism feels like a generically useful concept so implementing it in a way that we can reuse it for other similar gates that require manual admin action felt right, vs tying it to a specific component.


I guess another rejected alternative was to just use the apiserver's data about what apis are in use, and block upgrades if removed apis are in use.  It was rejected because that information isn't sufficient to make an informed choice (just because you used the api last week doesn't mean you care about it, and just because you didn't use it in the last week doesn't mean you don't have a tool or workload that is going to try to use it tomorrow)

Reply:
@jottofar Do you mind copying and pasting the two alternatives that Ben mentioned here and then I think we're good to approve and merge. I don't see anything else outstanding that would block us from merging this.

Reply:
Done